resolve.c

```diff
@@ -62,68 +70,92 @@ char *resolveHostname(const char *addr)
+		// If not, we will try to re-resolve all known clients
+		if(onlynew && clients[i].new)
 		{
-			// Delete possibly already existing hostname pointer before storing new data
-			if(clients[clientID].name != NULL)
-			{
-				free(clients[clientID].name);
-				clients[clientID].name = NULL;
-			}
-
-			// Store client hostname
-			clients[clientID].name = strdup(hostname);
-			clients[clientID].new = false;
+			char *hostname = resolveHostname(clients[i].ip);
+
+			enable_thread_lock();
+
+			if(clients[i].name != NULL)
+				free(clients[i].name);
+
+			clients[i].name = hostname;
+			clients[i].new = false;
+
+			disable_thread_lock();
 		}
-		free(hostname);
 	}
 }
 
-// This routine is run *before* saving to the database (default interval is once per minute)
-// to account for new clients (and forward destinations)
-void resolveNewClients(void)
+// Resolve upstream destination host names
+void resolveForwardDestinations(bool onlynew)
 {
 	int i;
-	for(i = 0; i < counters.clients; i++)
+	for(i = 0; i < counters.forwarded; i++)
 	{
 		// Memory validation
-		validate_access("clients", i, true, __LINE__, __FUNCTION__, __FILE__);
+		validate_access("forwarded", i, true, __LINE__, __FUNCTION__, __FILE__);
 
-		// Only try to resolve new clients
-		// Note that it can happen that we are not able to find hostnames but we don't
-		// want to try to resolve them every minute in this case.
-		if(clients[i].new)
+		// If onlynew flag is set, we will only resolve new upstream destinations
+		// If not, we will try to re-resolve all known upstream destinations
+		if(onlynew && forwarded[i].new)
 		{
-			clients[i].name = resolveHostname(clients[i].ip);
-			clients[i].new = false;
+			char *hostname = resolveHostname(forwarded[i].ip);
+
+			enable_thread_lock();
+
+			if(forwarded[i].name != NULL)
+				free(forwarded[i].name);
+
+			forwarded[i].name = hostname;
+			forwarded[i].new = false;
+
+			disable_thread_lock();
 		}
 	}
-	for(i = 0; i < counters.forwarded; i++)
+}
+
+void *DNSclient_thread(void *val)
+{
+	// Set thread name
+	prctl(PR_SET_NAME, "DNS client", 0, 0, 0);
+
+	while(!killed)
 	{
-		// Memory validation
-		validate_access("forwarded", i, true, __LINE__, __FUNCTION__, __FILE__);
+		// Run every minute to resolve only new clients and upstream servers
+		if(time(NULL) % RESOLVE_INTERVAL == 0)
+		{
+			// Try to resolve new client host names (onlynew=true)
+			resolveClients(true);
+			// Try to resolve new upstream destination host names (onlynew=true)
+			resolveForwardDestinations(true);
+			// Prevent immediate re-run of this routine
+			sleepms(500);
+		}
 
-		// Only try to resolve new forward destinations
-		if(forwarded[i].new)
+		// Run every hour to update possibly changed client host names
+		if(time(NULL) % RERESOLVE_INTERVAL == 0)
 		{
-			forwarded[i].name = resolveHostname(forwarded[i].ip);
-			forwarded[i].new = false;
+			// Try to resolve all client host names (onlynew=false)
+			resolveClients(false);
```

Comment:
Only clients are re-resolved here, not upstreams like in `>reresolve`.

Reply:
I'm still not a big fan of resolving the upstream names frequently, but I agree that it should be consistent, so I'll add it.

Reply:
I'm not totally for or against it, so if you think it is fine without it, then it can be removed.

Reply:
We can leave it in, but should keep in mind that this changes how FTL did it before (although there will probably no effect at all) and that there will be a few more PTR queries.